src/update-utmp/update-utmp.c

```diff
@@ -132,6 +135,15 @@ static int on_reboot(Context *c) {
          * utmp_put_reboot() will then fix to the current time */
         t = get_startup_time(c);
 
+        /* If RTC is in localtime, apply the delta to have 
+	 * number of seconds from EPOCH */
+        if (clock_is_localtime(NULL) > 0) {
+                assert_se(clock_gettime(CLOCK_REALTIME, &ts) == 0);
+                assert_se(tm = localtime(&ts.tv_sec));
+
+                t = t - tm->tm_gmtoff * USEC_PER_SEC;
+        }       
```

Comment:
I am sorry, what? You are suggesting that utmp stores time in local time instead of UTC time? What makes you think so?

Reply:
Hello,

If I understand well, if the RTC is not in UTC, the call to dual_timestamp_get(&userspace_timestamp); sets the UserSpaceTimestamp in localtime, before the set of the timezone with r = clock_set_timezone(&min);
---
ts->realtime = now(CLOCK_REALTIME);
---
- This is visible with VM with RTC in localtime of hypervisor.
I know that it's not recommended.

Thanks

Reply:
No. dual_timestamp_get() takes the CLOCK_REALTIME/CLOCK_MONOTONIC clocks. And CLOCK_REALTIME is UNIX time, i.e. UTC.

But I think I grok what you are trying to say: the "userspace" timestamp is taken during very early runtime, and thus possibly skewed on systems that run in local RTC time, since at that point the kernel might have initialized the system clock from RTC believing it was in UTC, and we only apply the clock jump to fix that up a bit later than where we got the timestamps.

Since what we store in utmp is directly the userspace timestamp we thus might store the uncorrected time, and your code tries to adjust that again.

I am not sure I like the approach much though, given that things will fall apart if the rtc-local boolean setting is toggled in the mean time.

a better approach would probably be to make use of the monotonic clock here. after all pid1 records both CLOCK_REALTIME *and* CLOCK_MONOTONIC early on. hence, when it comes to writing out the utmp record, maybe base it on the CLOCK_MONOTONIC timestamp: use the current CLOCK_REALTIME timestamp and subtract the time that passed since the CLOCK_MONOTONIC timestamp. The map_clock_usec() helper does this mapping for you.

The monotonic clock is exported in the UserspaceTimestampMonotonic dbus property already, update-utmp.c would just have to query that.

Reply:
Thanks Lennart,

Should we fix the value we set in utmp or the UserSpaceTimestamp  when we see that the RTC is not in UTC and we adjust the timezone ? 
What should be the reference in UserSpaceTimestamp ?

Thank you !

Reply:
don't check what the rtc is in. do it uconditionally. the rtc might also be off, and got fixed later.

Reply:
https://github.com/caoxia2008cxx/systemd/pull/2/commits/21ccfd4c7a6e3541ef2a60e75002052d9c469021
@poettering can this commit work?and I can't  put this commit to bwelterl:RTC_not_in_UTC branch.